### PR TITLE
[Matthew Phua] iP

### PR DESCRIPTION
**Pull request by matthewphua**

### CODE REVIEW

File: src/main/java/Duke.java [hunk 1]
```diff
@@ -1,10 +1,108 @@
...
+        Scanner in = new Scanner(System.in);
+        String line;
+
+        System.out.println(HORIZONTAL_LINE + " Hello! I'm Matthew\n" + " What can I do for you?\n" + HORIZONTAL_LINE);
+
+        while (true) {
+            line = in.nextLine();
+            if (line.equals("bye")) {
+                break;
+            } else if (line.equals("list")){
```
Comment (pipipipi2002) [inline src/main/java/Duke.java]: no space before the last '{' on this line

File: src/main/java/Event.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
+public class Event extends Task {
+
+    protected String at;
+
+    public Event(String description, String at) {
+        super(description);
+        this.at = at;
+    }
+
+    public String toString() {
```
Comment (pipipipi2002) [inline src/main/java/Event.java:29]: Consider writting `@Override` for `toString()` methods. Same for the other Classes.

**Review by pipipipi2002 — COMMENTED** [review]
Amazing job, i am very impressed by the code cleanliness and use of OOP :)

File: src/main/java/Duke.java [hunk 3]
```diff
@@ -1,10 +1,108 @@
...
+    private static Task[] inputs = new Task[100];
+    private static int index = 0;
+
+    /**
+     * Returns a string with formatted user inputs.
+     *
+     * @param inputs list of user input.
+     * @return Formatted string to print
+     */
+    public static String formatList(Task[] inputs) {
```
Comment (Franky4566) [inline src/main/java/Duke.java]: No comments for all functions. Can consider adding based on the java coding standards so that its easier to read & understand your code

File: src/main/java/Duke.java [hunk 4]
```diff
@@ -1,10 +1,108 @@
...
+
+    private static void mark(String line) {
+        int input = Integer.valueOf(line.replace("mark", "").strip());
+        Task task = inputs[input - 1];
+        task.markAsDone();
+        System.out.println(HORIZONTAL_LINE + "Nice! I've marked this task as done: \n  "
+                + task + "\n"+ HORIZONTAL_LINE);
+    }
+
+    private static void createTodo(String line){
```
Comment (Franky4566) [inline src/main/java/Duke.java]: add spacing before  bracket {

File: src/main/java/Duke.java [hunk 5]
```diff
@@ -1,10 +1,108 @@
...
+        String description = line.replace("todo", "").strip();
+        Todo todo = new Todo(description);
+        inputs[index] = todo;
+        index++;
+        System.out.println(HORIZONTAL_LINE + ADD_CONFIRMATION_MESSAGE + "\n  " + todo
+                + "\nNow you have "+ index + " tasks in the list.\n" + HORIZONTAL_LINE);
+    }
+
+    private static void createEvent(String line){
```
Comment (Franky4566) [inline src/main/java/Duke.java]: spacing before bracket {

**Review by Franky4566 — COMMENTED** [review]
Good job! just some minor issues

File: src/main/java/Todo.java [hunk 6]
```diff
@@ -0,0 +1,11 @@
+public class Todo extends Task {
+
+    public Todo(String description) {
+        super(description);
+    }
+
+    public String toString() {
+        return "[T] " + super.toString();
+    }
+
```
Comment (pipipipi2002) [inline src/main/java/Todo.java]: Add `@Override`

File: src/main/java/Task.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+     *
+     * @return X if task is done, " " if not done
+     */
+    public String getStatusIcon() {
+        return (isDone ? "X" : " "); // mark done task with X
+    }
+
+    public String toString() {
+        return "["+ getStatusIcon()+ "] " + this.description;
+    }
```
Comment (pipipipi2002) [inline src/main/java/Task.java:31]: Add `@Override`

File: src/main/java/Task.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    public void markAsNotDone() {
+        this.isDone = false;
+    }
+
+    /**
+     * Returns description for a task
+     *
+     * @return the description for a task
+     */
+    public String getDescription(){
```
Comment (pipipipi2002) [inline src/main/java/Task.java]: Add space before `{`

File: src/main/java/Duke.java [hunk 9]
```diff
@@ -1,10 +1,108 @@
...
+            if (inputs[i] == null) {
+                break;
+            }
+            formattedString += (i+1) + "." + inputs[i].toString() + "\n";
+        }
+        formattedString += HORIZONTAL_LINE;
+        return formattedString;
+    }
+
+    private static void unmark(String line){
```
Comment (pipipipi2002) [inline src/main/java/Duke.java]: add space before `{`

**Review by pipipipi2002 — COMMENTED** [review]
Great Job!

File: src/main/java/Duke.java [hunk 10]
```diff
@@ -1,10 +1,108 @@
+import java.util.Scanner;
+
 public class Duke {
+    private static final String HORIZONTAL_LINE = "____________________________________________________________\n";
```
Comment (malcolmang) [inline src/main/java/Duke.java]: Good use of static constants here!

File: src/main/java/Duke.java [hunk 11]
```diff
@@ -1,10 +1,108 @@
...
+
     public static void main(String[] args) {
-        String logo = " ____        _        \n"
-                + "|  _ \\ _   _| | _____ \n"
-                + "| | | | | | | |/ / _ \\\n"
-                + "| |_| | |_| |   <  __/\n"
-                + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+        Scanner in = new Scanner(System.in);
+        String line;
```
Comment (malcolmang) [inline src/main/java/Duke.java]: "Variables should be initialized where they are declared and they should be declared in the smallest scope possible." 
In this case this String `line` could be declared and initialized in the while loop below, in a slightly smaller scope.

File: src/main/java/Duke.java [hunk 12]
```diff
@@ -1,10 +1,108 @@
...
+            } else if (line.startsWith("unmark")) {
+                unmark(line);
+            } else if (line.startsWith("todo")) {
+                createTodo(line);
+            } else if (line.startsWith("event")) {
+                createEvent(line);
+            } else if (line.startsWith("deadline")) {
+                createDeadline(line);
+            }  else {
+                createTask(line);
```
Comment (malcolmang) [inline src/main/java/Duke.java]: In this case the Task and Todo classes seem functionally the same (they both only have a description, but the Todo class contains an extra marker `[T]`). You may want to consider making Todo tasks the default and remove the ability to instance the `Task` superclass with `abstract` in front of the class declaration of `Task`.

File: src/main/java/Duke.java [hunk 13]
```diff
@@ -1,10 +1,108 @@
...
+    }
+
+    private static void createEvent(String line){
+        String removedCommand = line.replace("event", "");
+        String description = removedCommand.substring(0, removedCommand.indexOf("/at")).strip();
+        String date = removedCommand.substring(removedCommand.indexOf("/at") + "/at".length()).strip();
+        Event event = new Event(description, date);
+        inputs[index] = event;
+        index++;
+        System.out.println(HORIZONTAL_LINE + ADD_CONFIRMATION_MESSAGE + "\n  " + event
```
Comment (malcolmang) [inline src/main/java/Duke.java]: This pattern of `HORIZONTAL_LINE + {item} + "\n" + HORIZONTAL_LINE` appears multiple times in your code. Maybe extracting that into a reusable method might make the code more readable.

For example:
```
void print(String printString) {
    System.out.println(HORIZONTAL_LINE + printString + "\n" + HORIZONTAL_LINE);
}
```

That way you could just call here 
```
print(ADD_CONFIRMATION_MESSAGE + "\n " + event 
        + "\nNow you have "+ index + " tasks in the list.");
```
and reuse it as needed.

File: src/main/java/Duke.java (hunk 3 above)
Comment (malcolmang) [inline src/main/java/Duke.java]: Overall, your main Duke class is doing a lot of heavy lifting. You may want to consider creating a separate class to deal with formatting and getting user input(UI class), and maybe even one more for just creating and managing Tasks (TaskManager class), and make the Duke class a little cleaner and easier to follow.

File: src/main/java/Duke.java [hunk 14]
```diff
@@ -1,10 +1,108 @@
...
+                + "\nNow you have "+ index + " tasks in the list.\n" + HORIZONTAL_LINE);
+    }
+
+    private static void createDeadline(String line) {
+        String removedCommand = line.replace("deadline", "");
+        String description = removedCommand.substring(0, removedCommand.indexOf("/by")).strip();
```
Comment (malcolmang) [inline src/main/java/Duke.java]: This could be extracted into a method and reused (a similar pattern is seen in the createEvent method, where the only difference is the string passed to `.indexOf()`.

**Review by malcolmang — COMMENTED** [review]
Overall quite a good implementation. Great job! The code works but could be a little easier to read if given a bit more structure.

File: src/main/java/DukeManager.java [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+            return;
+        }
+        inputs[index] = todo;
+        index++;
+        print(ADD_CONFIRMATION_MESSAGE + "\n  " + todo + "\nNow you have "+ index + " tasks in the list.");
+    }
+
+    public static void createEvent(String line) {
+        String removedCommand = line.replace("event", "");
+        int timeIndex = removedCommand.indexOf("/at");
```
Comment (hughjazzman) [inline src/main/java/DukeManager.java]: Could `"/at"` be a magic literal? Similarly for `"/by"`?

https://nus-cs2113-ay2223s1.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-magic-numbers

File: src/main/java/DukeManager.java [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+     * @param inputs list of user input.
+     * @return Formatted string to print
+     */
+    public static String formatList(Task[] inputs) {
+        String formattedString = HORIZONTAL_LINE;
+        for (int i = 0; i < inputs.length; i++) {
+            if (inputs[i] == null) {
+                break;
+            }
+            formattedString += (i+1) + "." + inputs[i].toString() + "\n";
```
Comment (hughjazzman) [inline src/main/java/DukeManager.java]: Perhaps the operators can be separated by whitespace?
```suggestion
            formattedString += (i + 1) + "." + inputs[i].toString() + "\n";
```

> Operators should be surrounded by a space character.

https://se-education.org/guides/conventions/java/index.html#layout

**Review by hughjazzman — COMMENTED** [review]
Nice code

File: src/main/java/DukeManager.java [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+            print("Something went wrong when trying to save: " + e.getMessage());
+            return;
+        }
+        String pathName = "./data/duke.txt";
+        try {
+            FileWriter fw = new FileWriter(pathName);
+            fw.write(saveList());
+            fw.close();
+        } catch (IOException e) {
+            print("☹ OOPS!!! Something went wrong when trying to save: " + e.getMessage());
```
Comment (RiaVora) [inline src/main/java/DukeManager.java]: Since you use this in multiple areas, maybe consider making this error message a constant you can easily reference!

File: src/main/java/DukeManager.java [hunk 18]
```diff
@@ -0,0 +1,238 @@
...
+            return;
+        }
+    }
+
+    public static void load() {
+        if (Files.exists(directoryPath) && Files.exists(filePath)) {
```
Comment (RiaVora) [inline src/main/java/DukeManager.java]: if the file path exists, then the directoryPath has to exist, so you can probably shorten this to just "Files.exists(filePath)"!!

File: src/main/java/DukeManager.java [hunk 19]
```diff
@@ -0,0 +1,238 @@
...
+            File f = new File(String.valueOf(filePath)); // create a File for the given file path
+            Scanner s = null;
+            try {
```
Comment (RiaVora) [inline src/main/java/DukeManager.java]: consider putting all of this code in a try catch, since your error message is fairly similar and you can catch two different types of exceptions in one try catch :)

File: text-ui-test/EXPECTED.TXT [hunk 20]
```diff
@@ -1,7 +1,5 @@
-Hello from
- ____        _        
-|  _ \ _   _| | _____ 
-| | | | | | | |/ / _ \
-| |_| | |_| |   <  __/
-|____/ \__,_|_|\_\___|
+____________________________________________________________
```
Comment (RiaVora) [inline text-ui-test/EXPECTED.TXT:1]: I like how you added your own quirks to the Duke app!

File: src/main/java/DukeException.java [hunk 21]
```diff
@@ -0,0 +1,2 @@
+public class DukeException extends Exception {
```
Comment (RiaVora) [inline src/main/java/DukeException.java:1]: Consider making a constructor with an error message, so you can preserve different errors and print their messages when making and returning DukeExceptions!

File: src/main/java/DukeManager.java [hunk 22]
```diff
@@ -1,35 +1,145 @@
...
+import java.nio.file.Paths;
+import java.nio.file.Files;
+import java.util.Scanner;
+
 public class DukeManager {
     private static final String HORIZONTAL_LINE = "____________________________________________________________\n";
     private static final String ADD_CONFIRMATION_MESSAGE = "Got it. I've added this task:";
-    private static Task[] inputs = new Task[100];
-    private static int index = 0;
+    private static final Path directoryPath = Paths.get(".","data");
```
Comment (hughjazzman) [inline src/main/java/DukeManager.java]: Awesome use of `Paths`!

File: src/main/java/DukeManager.java [hunk 23]
```diff
@@ -1,35 +1,145 @@
...
+    public static String formatList() {
         String formattedString = HORIZONTAL_LINE;
-        for (int i = 0; i < inputs.length; i++) {
-            if (inputs[i] == null) {
+        for (int i = 0; i < inputs.size(); i++) {
+            if (inputs.get(i) == null) {
                 break;
             }
-            formattedString += (i+1) + "." + inputs[i].toString() + "\n";
+            formattedString += (i+1) + "." + inputs.get(i).toString() + "\n";
```
Comment (hughjazzman) [inline src/main/java/DukeManager.java]: Perhaps the operator should be separated by space?
```suggestion
            formattedString += (i + 1) + "." + inputs.get(i).toString() + "\n";
```
> Operators should be surrounded by a space character.

https://se-education.org/guides/conventions/java/index.html#layout

File: src/main/java/DukeManager.java [hunk 24]
```diff
@@ -1,35 +1,145 @@
...
+    }
+
+    public static void save() {
+        try {
+            checkExists();
+        } catch (IOException e) {
+            print("Something went wrong when trying to save: " + e.getMessage());
+            return;
+        }
+        String pathName = "./data/duke.txt";
```
Comment (hughjazzman) [inline src/main/java/DukeManager.java]: Since you're already using `Path` above, there could be some consistency here instead of `String`?

https://nus-cs2113-ay2223s1.github.io/website/se-book-adapted/chapters/codeQuality.html#do-not-trip-up-reader

File: src/main/java/DukeManager.java [hunk 25]
```diff
@@ -1,35 +1,145 @@
...
+            } catch (FileNotFoundException e) {
+                print("☹ OOPS!!! Something went wrong when trying to load: " + e.getMessage());
+            }
+            while (s.hasNext()) {
+                String[] line = s.nextLine().split("\\|");
+                Task newTask = null;
+                if (line[0].equals("T")) {
+                    try {
+                        newTask = new Todo(line[2]);
+                    } catch (DukeException e) {
```
Comment (hughjazzman) [inline src/main/java/DukeManager.java]: Some repetitive code here you can consider restructuring?

https://nus-cs2113-ay2223s1.github.io/website/se-book-adapted/chapters/codeQuality.html#minimise-code-duplication

**Review by hughjazzman — COMMENTED** [review]
Great work!
